### PR TITLE
feat: Add `__tools__` optional param for function pipes

### PR DESCRIPTION
**Pull request by michaelpoluektov**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests for validating the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To cleary categorize this pull request, prefix the pull request title, using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- Add optional `__tools__` param to function pipes.

### Added

- Function pipes accept optional `__tools__: dict[str, dict]` param for tool calling.

The `__tools__` dict maps tool names to a dict with keys `"toolkit_id"`, `"callable"`, `"spec"`, `pydantic_model`, `"file_handler"` and `"citation"` where `"file_handler"` and `"citation"` are booleans for whether or not the tool handles files and has a citation object (mostly useful for the tool prepending thing) and `"spec"` is an OpenAI compatible "function" spec for the tool.

# Testing

Tested with Ollama, OpenAI and an OpenAI proxy function, with the flags enabled and disabled.

### CODE REVIEW

**Comment by michaelpoluektov** [comment]
My bad: this needs to be on a per-model basis rather than as an environment variable. Converted to draft for now

**Comment by michaelpoluektov** [comment]
Sample function to make use of this:

```python
"""
title: OpenAI ReAct
author: Michael Poluektov
author_url: https://github.com/michaelpoluektov
description: OpenAI ReAct
version: 0.1.0
"""

import os
from typing import Generator, Iterator

from pydantic import BaseModel, Field
from openai import OpenAI
from langchain_openai import ChatOpenAI
from langchain_core.tools import StructuredTool
from langgraph.prebuilt import create_react_agent

BAD_NAMES = ["202", "13", "3.5", "preview"]


class Pipe:
    class Valves(BaseModel):
        OPENAI_BASE_URL: str = Field(
            default="https://api.openai.com/v1",
            description="Base URL for OpenAI API endpoints",
        )
        OPENAI_API_KEY: str = Field(default="", description="OpenAI API key")
        TITLE_MODEL: str = Field(
            default="gpt-4o-mini", description="Model used to generate title"
        )

    def __init__(self):
        self.type = "manifold"
        self.valves = self.Valves(
            **{k: os.getenv(k, v.default) for k, v in self.Valves.model_fields.items()}
        )
        print(f"{self.valves=}")

    def pipes(self) -> list[dict[str, str]]:
        if not self.valves.OPENAI_API_KEY:
            return [{"id": "Error: OPENAI_API_KEY is not set"}]
        try:
            openai = OpenAI(
                api_key=self.valves.OPENAI_API_KEY, base_url=self.valves.OPENAI_BASE_URL
            )
            models = [
                m.id
                for m in openai.models.list().data
                if "gpt" in m.id and not any(n in m.id for n in BAD_NAMES)
            ]
            return [{"id": m, "name": f"OpenAI/{m}"} for m in models]
        except Exception as e:
            return [{"id": "error", "name": f"Could not fetch models: {e}"}]

    async def pipe(
        self,
        body: dict,
        __user__: dict,
        __task__: str | None,
        __tools__: dict[str, dict] | None,
    ) -> str | Generator | Iterator:
        print(__task__)
        print(f"{__tools__=}")
        if __task__ == "function_calling":
            return ""
        if not self.valves.OPENAI_API_KEY or not self.valves.OPENAI_BASE_URL:
            raise Exception("Error: OPENAI_API_KEY or OPENAI_BASE_URL is not set")

        if __task__ == "title_generation":
            model_id = self.valves.TITLE_MODEL
        else:
            model_id = body["model"][body["model"].rfind(".") + 1 :]

        model = ChatOpenAI(
            model=model_id,
            base_url=self.valves.OPENAI_BASE_URL,
            api_key=self.valves.OPENAI_API_KEY,  # type: ignore
            temperature=0,
        )

        if __task__ == "title_generation":
            out = model.invoke(body["messages"]).content
            assert isinstance(out, str)
            return out

        if __tools__ is not None:
            tools = []
            for key, value in __tools__.items():
                tools.append(
                    StructuredTool(
                        func=None,
                        name=key,
                        coroutine=value["callable"],
                        args_schema=value["pydantic_model"],
                        description=value["spec"]["description"],
                    )
                )
        else:
            tools = []
        graph = create_react_agent(model, tools=tools)  # type: ignore
        print(f'{body["messages"] = }')
        inputs = {"messages": body["messages"]}
        final_result = None
        async for i in graph.astream(inputs, stream_mode="values"):
            i["messages"][-1].pretty_print()
            final_result = i
        if final_result is not None:
            final_result["messages"][-1].pretty_print()
            return final_result["messages"][-1].content
        return "Nope."
```

**Review by Peter-De-Ath — APPROVED** [review]

File: backend/apps/ollama/main.py [hunk 1]
```diff
@@ -732,11 +732,10 @@ async def generate_chat_completion(
 ):
     log.debug(f"{form_data.model_dump_json(exclude_none=True).encode()}=")
 
-    payload = {
```
Comment (tjbck) [inline backend/apps/ollama/main.py:735]: Let's put `files` and `tool_ids` to metadata

**Comment by tjbck** [comment]
The rest LGTM, Thanks!

File: backend/apps/ollama/main.py (hunk 1 above)
Comment (michaelpoluektov) [reply backend/apps/ollama/main.py:736]: Done

**Comment by tjbck** [comment]
LGTM, Thanks!

**Pull request merged**